### ISSUE TITLE
failure when using google_project_service

### ISSUE THREAD

**Issue by clembo590** [hashicorp/terraform-provider-google#8214]
Using this version
```
  "registry.terraform.io/hashicorp/google": {
    "hash": "h1:+r2711CMnrdJgVLo9e034qPaVJmmEqYdt2JBycP6uaU=",
    "version": "3.52.0"
  },
```

I am first describing a scenario that works, and then a scenario that does not work. (to show the error)
# COMMON TO BOTH SCENARIO
## set up variables

```
export TF_VAR_org_id=YOUR_ORG_ID
export TF_VAR_billing_account=YOUR_BILLING_ACCOUNT_ID
export TF_ADMIN=${USER}-terraform-admin
export TF_CREDS=~/.config/gcloud/${USER}-terraform-admin.json
export GOOGLE_APPLICATION_CREDENTIALS=${TF_CREDS}
export GOOGLE_PROJECT=${TF_ADMIN}
```

# SCENARIO THAT WORKS CURRENTLY:
## step 1
create a main.tf
```
variable "project_name" {}
variable "billing_account" {}
variable "org_id" {}
variable "region" {}

provider "google" {
  region = var.region
}

resource "random_id" "id" {
  byte_length = 4
  prefix      = var.project_name
}

resource "google_project" "project" {
  name            = var.project_name
  project_id      = random_id.id.hex
  billing_account = var.billing_account
  org_id          = var.org_id
}

resource "google_project_service" "service" {
  for_each = toset([
    "compute.googleapis.com"
  ])

  service = each.key

  project            = google_project.project.project_id
  disable_on_destroy = false
}

output "project_id" {
  value = google_project.project.project_id
}
```
`terraform apply -auto-approve `
--> **SUCCESS**

## step 2
add following to the end of the main.tf

```
data "google_compute_zones" "available" {
  project = google_project.project.project_id
}

resource "google_compute_instance" "default" {
  project      = google_project.project.project_id
  zone         = data.google_compute_zones.available.names[0]
  name         = "tf-compute-1"
  machine_type = "f1-micro"

  boot_disk {
    initialize_params {
      image = "ubuntu-1604-xenial-v20170328"
    }
  }

  network_interface {
    network = "default"
    access_config {}
  }

  depends_on = [google_project_service.service]
}

output "instance_id" {
  value = google_compute_instance.default.self_link
}
```
`terraform apply -auto-approve `
--> **SUCCESS**

`terraform destroy -auto-approve `
--> **SUCCESS**

# SCENARIO THAT DOES NOT WORK:
perfom the same thing in only **ONE** step (your already have the 'full' main.tf file)
`terraform apply -auto-approve `
--> **FAILURE**

```
google_project_service.service["compute.googleapis.com"]: Creation complete after 1m42s [id=clement-test-compute0316d75e/compute.googleapis.com]

Error: googleapi: Error 403: Compute Engine API has not been used in project 593862021661 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/compute.googleapis.com/overview?project=593862021661 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry., accessNotConfigured
```

I have run these 2 scenarios 10 times:
when ran in 2 steps it always succeed.
when ran in 1 step it always fails.

**Comment by edwardmedia** [comment]
@clembo590 this is likely a time issue. When you run all code together, it starts to create instances while the services are not ready yet. Can you add [sleep](https://registry.terraform.io/providers/hashicorp/time/latest/docs/resources/sleep) logic between resources to see if that helps?

**Comment by clembo590** [comment]
@edwardmedia 
Thanks for you answer. that is also what I thought when I realized that creating the infrastucture in 2 steps was working.
So it s why I had the idea of adding a "sleep". (I did that before I posted the bug here).
Just to make sure I retried and it still fails
here is the main.tf

```
variable "project_name" {}
variable "billing_account" {}
variable "org_id" {}
variable "region" {}

provider "google" {
  region = var.region
}

resource "random_id" "id" {
  byte_length = 4
  prefix      = var.project_name
}

resource "google_project" "project" {
  name            = var.project_name
  project_id      = random_id.id.hex
  billing_account = var.billing_account
  org_id          = var.org_id
}

resource "google_project_service" "service" {
  for_each = toset([
    "compute.googleapis.com"
  ])

  service = each.key

  project            = google_project.project.project_id
  disable_on_destroy = false
}

output "project_id" {
  value = google_project.project.project_id
}


# BELOW IS ABOUT CREATING A COMPUTE INSTANCE

resource "time_sleep" "wait_180_seconds" {
  depends_on = [google_project_service.service]
  create_duration = "180s"
}

data "google_compute_zones" "available" {
  project = google_project.project.project_id
}

resource "google_compute_instance" "default" {
  project      = google_project.project.project_id
  zone         = data.google_compute_zones.available.names[0]
  name         = "tf-compute-1"
  machine_type = "f1-micro"

  boot_disk {
    initialize_params {
      image = "ubuntu-1604-xenial-v20170328"
    }
  }

  network_interface {
    network = "default"
    access_config {}
  }

  depends_on = [google_project_service.service,time_sleep.wait_180_seconds]
}

output "instance_id" {
  value = google_compute_instance.default.self_link
}

```

I also generated the graph

![Screenshot 2021-01-14 at 20 03 02](https://user-images.githubusercontent.com/10899818/104636388-94bcef00-56a3-11eb-8f34-91424deafb9a.png)


It s failing with same error message:
Error: googleapi: Error 403: Compute Engine API has not been used in project 96126660076 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/compute.googleapis.com/overview?project=96126660076 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry., accessNotConfigured

**Comment by edwardmedia** [comment]
@clembo590 can you try by giving **sleep** a bit longer?

**Comment by clembo590** [comment]
@edwardmedia 
I put I 10 minutes sleep. Just to make sure --> it s still failing


```
time_sleep.wait_600_seconds: Creation complete after 10m0s [id=2021-01-14T21:21:25Z]

Error: googleapi: Error 403: Compute Engine API has not been used in project 122036389144 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/compute.googleapis.com/overview?project=122036389144 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry., accessNotConfigured

```

This is a strange issue....

**Comment by edwardmedia** [comment]
@clembo590 you have verified that once you separate the config into different pieces and execute them separately, it is all good. No code change needed. Am I right?  The difference between these two ways is the timing, which matches the error you received. There were many reported `issue` about enabling services like yours. Here is one [example](https://github.com/hashicorp/terraform-provider-google/issues/6377) Let me know if the case is different.

**Comment by clembo590** [comment]
@edwardmedia to answer your question:
_you have verified that once you separate the config into different pieces and execute them separately, it is all good. No code change needed. Am I right?_
--> yes you are right.

I have put a "time_sleep" at every step and in the end I managed to make it work with "only" one "terraform apply".
--> this means that indeed the timing is the issue.
--> this means that there is definitly a bug, Am I right ?

**Comment by edwardmedia** [comment]
@clembo590 the provider plugin relies on the individual resource that reports whether the provision is complete. For most cases, that is accurate. But for small number of resources, that is not. Unfortunately some services are among of them. This is something that is beyond what the provider can do. Does this make sense to you?

**Comment by clembo590** [comment]
@edwardmedia 
What you say makes sense.  in my opinion there is only two ways to proceed further:
Do you consider it fails because I wrote something wrong ?
Then it s my fault (sorry about that). Or I did not do anything wrong and then I think the issue should be investigated/resolved. (or just update documentation so that we know that we must put a sleep timer when we create XXXX resource with this provider) .



The bug description I gave cannot be more clear.
It also cannot be more simple: create a project, enable GCE on the project, create an instance in GCE. (simple is it not?)

**Comment by edwardmedia** [comment]
@clembo590 adding `sleep` in Terraform config is not uncommon when you provision a system. If you encounter problems on individual resources, that is something we can help. Closing this issue then

**Comment by clembo590** [comment]
@edwardmedia 
Hello. Good news is that this bug has been fixed in version 3.54.
Can you please remove the label "question" and replace it by "bug". 
thank you.

**Comment by rileykarson** [comment]
Brief update, leaving this thread locked: This remains an issue. Service activation in GCP APIs is not strongly consistent, and even with multiple success checks on Terraform's side (LRO finishing + the service appearing activated in a service List call) that doesn't guarantee it's actually usable, particularly for just-created projects. There's some Google-internal discussion at b/228371710.
